File: built-in plugins dsp/mb-fm-epiano.cpp
/**
 * MB FM E-Piano
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : FM electric piano tines
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_EPIANO_H
#define MB_FM_EPIANO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmEpiano : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-epiano";
    static constexpr const char* PLUGIN_NAME    = "MB FM E-Piano";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float modIndex = 2.5f;  // range [0, 15]
    float brightness = 0.6f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbFmEpiano() = default;
    ~MbFmEpiano() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.modIndex = std::clamp(params.modIndex, 0f, 15f);
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM E-Piano
        return input;
    }
};

#endif // MB_FM_EPIANO_H

File: built-in plugins dsp/mb-fm-epiano.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFmEpianoPlugin: PluginDefinition = { id: 'mb-fm-epiano', slug: 'mb-fm-epiano', name: 'MB FM E-Piano', category: 'instrument', type: 'fm', version: '1.0.0', description: 'FM electric piano tines', author: 'Max Booster', grade: 'A', oscillators: [{ type: 'sine', detune: 0, gain: 0.7 }], envelope: { attack: 0.001, decay: 0.8, sustain: 0.3, release: 0.5 }, parameters: [{ id: 'modIndex', name: 'Mod Index', type: 'float', defaultValue: 2.5, minValue: 0, maxValue: 15, automatable: true }, { id: 'brightness', name: 'Brightness', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { modIndex: 2.5, brightness: 0.6, volume: 0.8 } };

export default MbFmEpianoPlugin;
